test/Tests/src/Data/Array_Proxy_Spec.enso

```diff
@@ -37,7 +37,17 @@ spec =
             v2 = Vector.from_polyglot_array (Array_Proxy.from_proxy_object (Proxy_Object.Value 5))
             v2.should_equal [0, 10, 20, 30, 40]
 
-        Test.specify "should check the callback type validity at construction" <|
+        Test.specify "should check validity at construction" <|
             Test.expect_panic_with (Array_Proxy.new 0 0) Type_Error_Data
+            Test.expect_panic_with (Array_Proxy.new -1 (x->x)) Illegal_Argument_Error_Data
+
+        Test.specify "should pretend to just be an Array" <|
+            proxy = Array_Proxy.new 3 (ix -> ix + 10)
+            ok_match = case proxy of
+                _ : Vector.Vector -> "Vector"
+                _ : Array -> "Array"
+                _ -> "other"
+            ok_match . should_equal "Array"
+            Meta.get_qualified_type_name proxy . should_equal "Standard.Base.Data.Array.Array"
```

Comment:
This feels a bit odd even though this is consistent with `getType`.

Reply:
You mean the `qualified_type_name`? I'm ok with changing that if you think we should.

My main point is to ensure that it is matched by `_ : Array` - as it definitely should be the same way as `ArrayList` is. We should probably add tests for `ArrayOverBuffer` and `ArraySlice` too - looking at the impl of `CatchTypeBranchNode` I'm afraid currently they may not be matched this way.

But as for the qualified type name - I don't know, I guess we can also consider it returning `[...].Array_Proxy` instead. Do you think that would be better? I guess it should just return something 'better' than just `null` :)

Reply:
btw it would continue to match on `_ : Array` never mind the type (it looks at `hasArrayElements` message.
But if we stick to `Standard.Base.Data.Array.Array` type, `_ : Array_Proxy` won't match IIRC.

Reply:
Question is if we want `Array_Proxy` as a separate _type_ in the language. As @JaroslavTulach was suggesting, we may just have one `Array` type that has various underlying implementations. The `Array_Proxy` module can be treated like a special helper module for creating this particular type of `Array`. Or we can move these methods to `Array.create_proxy` or something like that.

Reply:
Looks like I missed your response. Jaroslav's suggestion sounds good. I think this can stay as-is.